src/cli_instance.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright (c) 2019-2022 Alibaba Cloud
+// Copyright (c) 2019-2022 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::{
+    path::PathBuf,
+    sync::{
+        mpsc::{Receiver, Sender},
+        Arc, RwLock,
+    },
+};
+
+use anyhow::{anyhow, Context, Result};
+use seccompiler::BpfProgram;
+use vmm_sys_util::eventfd::EventFd;
+
+use dragonball::{
+    api::v1::{
+        BlockDeviceConfigInfo, BootSourceConfig, InstanceInfo, VmmAction, VmmActionError, VmmData,
+        VmmRequest, VmmResponse,
+    },
+    vm::{CpuTopology, VmConfigInfo},
+};
+
+use crate::parser::DBSArgs;
+
+const DRAGONBALL_VERSION: &str = env!("CARGO_PKG_VERSION");
+const REQUEST_RETRY: u32 = 500;
+
+pub enum Request {
+    Sync(VmmAction),
+}
+
+pub struct CliInstance {
+    /// VMM instance info directly accessible from runtime
+    pub vmm_shared_info: Arc<RwLock<InstanceInfo>>,
+    pub to_vmm: Option<Sender<VmmRequest>>,
+    pub from_vmm: Option<Receiver<VmmResponse>>,
+    pub to_vmm_fd: EventFd,
+    pub seccomp: BpfProgram,
+}
+
+impl CliInstance {
+    pub fn new(id: &str) -> Self {
+        let vmm_shared_info = Arc::new(RwLock::new(InstanceInfo::new(
+            String::from(id),
+            DRAGONBALL_VERSION.to_string(),
+        )));
+
+        let to_vmm_fd = EventFd::new(libc::EFD_NONBLOCK)
+            .unwrap_or_else(|_| panic!("Failed to create eventfd for vmm {}", id));
+
+        CliInstance {
+            vmm_shared_info,
+            to_vmm: None,
+            from_vmm: None,
+            to_vmm_fd,
+            seccomp: vec![],
+        }
+    }
+
+    pub fn run_vmm_server(&mut self, args: DBSArgs) -> Result<()> {
+        // configuration
+        let vm_config = VmConfigInfo {
+            vcpu_count: args.create_args.vcpu,
+            max_vcpu_count: args.create_args.max_vcpu,
+            cpu_pm: args.create_args.cpu_pm.clone(),
+            cpu_topology: CpuTopology {
+                threads_per_core: args.create_args.cpu_topology.threads_per_core,
+                cores_per_die: args.create_args.cpu_topology.cores_per_die,
+                dies_per_socket: args.create_args.cpu_topology.dies_per_socket,
+                sockets: args.create_args.cpu_topology.sockets,
+            },
+            vpmu_feature: 0,
+            mem_type: args.create_args.mem_type.clone(),
+            mem_file_path: args.create_args.mem_file_path.clone(),
+            mem_size_mib: args.create_args.mem_size,
+            // as in crate `dragonball` serial_path will be assigned with a default value,
+            // we need a special token to enable the stdio console.
+            serial_path: Some("stdio".to_owned()),
```

Comment:
There should be some choices for user if they want to start a Dragonball VM but use vsock to connect to it

Reply:
:ok:

Reply:
```rust
            // as in crate `dragonball` serial_path will be assigned with a default value,
             // we need a special token to enable the stdio console.
             serial_path: Some(args.create_args.serial_path.clone()),
         };

         // check the existence of the serial path (rm it if exist)
         let serial_file = Path::new(&args.create_args.serial_path);
         if args.create_args.serial_path != *"stdio" && serial_file.exists() {
             std::fs::remove_file(serial_file).unwrap();
         }
```

Reply:
In `args.rs`:

```rust

   // The serial path used to communicate with VM
   #[clap(
       short,
       long,
       value_parser,
       default_value = "stdio",
       help = "The serial path used to communicate with VM",
       display_order = 2
   )]
   pub serial_path: String,
```

Reply:
please update the document